stdlib/regex.gr

```diff
@@ -0,0 +1,2946 @@
+  {
+    isPerlRegExp: config.isPerlRegExp,
+    caseSensitive: config.caseSensitive,
+    multiLine: multiLine,
+    groupNumber: config.groupNumber,
+    references: config.references,
+  }
+};
+
+let configGroupNumber = (config: RegExParserConfig) => unbox(config.groupNumber);
+
+let configIncGroupNumber = (config: RegExParserConfig) => {
+  config.groupNumber := unbox(config.groupNumber) + 1;
+  config
+};
+
+record RegExBuf {
+  input: String,
+  inputExploded: Array<Char>,
+  cursor: Box<Number>,
+  config: RegExParserConfig,
+};
+
+let makeRegExBuf = (s) => {
+  {input: s, inputExploded: String.explode(s), cursor: box(0), config: makeRegExParserConfig()}
+};
+
+let withConfig = (buf: RegExBuf, config: RegExParserConfig) => {
+  {input: buf.input, inputExploded: buf.inputExploded, cursor: buf.cursor, config: config}
+};
+
+// Parsing internals for recursive descent
+
+let parseErr = (buf: RegExBuf, msg: String, posShift) => {
+  "Invalid Regular Expression: " ++ msg ++ " (position " ++ toString(unbox(buf.cursor) + posShift) ++ ")"
+};
+
+let next = (buf: RegExBuf) => {
+  if (unbox(buf.cursor) >= Array.length(buf.inputExploded)) {
+    Err(parseErr(buf, "end of buffer reached", 0))
+  } else {
+    let ret = buf.inputExploded[unbox(buf.cursor)];
+    buf.cursor := unbox(buf.cursor) + 1;
+    Ok(ret)
+  }
+};
+
+let peek = (buf: RegExBuf) => {
+  if (unbox(buf.cursor) >= Array.length(buf.inputExploded)) {
+    Err(parseErr(buf, "end of buffer reached", 0))
+  } else {
+    Ok(buf.inputExploded[unbox(buf.cursor)])
+  }
+};
+
+let peekN = (buf: RegExBuf, n) => {
+  if (unbox(buf.cursor) + n >= Array.length(buf.inputExploded)) {
+    Err(parseErr(buf, "end of buffer reached", 0))
+  } else {
+    Ok(buf.inputExploded[unbox(buf.cursor) + n])
+  }
+};
+
+let eat = (buf: RegExBuf, char: Char) => {
+  if (unbox(buf.cursor) >= Array.length(buf.inputExploded)) {
+    Err(parseErr(buf, "end of buffer reached", 0))
+  } else {
+    let ret = buf.inputExploded[unbox(buf.cursor)];
+    if (ret == char) {
+      buf.cursor := unbox(buf.cursor) + 1;
+      Ok(ret)
+    } else {
+      Err(parseErr(buf, "Expected character '" ++ Char.toString(char) ++ ", but found character '" ++ Char.toString(ret) ++ "'", 0))
+    }
+  }
+};
+
+/**
+ * Checks if the given regex buffer is empty
+ * @param buf: RegExBuf - the buffer to check
+ * @returns Bool
+ */
+let more = (buf: RegExBuf) => {
+  unbox(buf.cursor) < Array.length(buf.inputExploded)
+};
+
+let moreN = (buf: RegExBuf, n) => {
+  unbox(buf.cursor) + n < Array.length(buf.inputExploded)
+};
+
+// END Parsing internals for recursive descent
+
+
+/*
+
+=================================
+REGEX RANGE DEFINITIONS AND UTILS
+=================================
+
+Based on https://github.com/racket/racket/blob/0a9c70e95a69743dd5d219a395e995be4a4bfd41/racket/src/regexp/common/range.rkt
+
+*/
+
+// [TODO] alias type RERange as List<(Number, Number)>
```

Comment:
Issue?

Reply:
#692

Reply:
Mind adding the issue number to the comment?